### PR TITLE
CORS-2950: Remove openshift-sdn as an install-time option

### PR DESCRIPTION
**Pull request by AnnaZivkovic**
The openshift-sdn CNI plug-in will no longer be an install-time option for newly installed 4.15+ clusters across installation options. Note that customer clusters currently using openshift-sdn that upgrade to 4.15 or 4.16 with openshift-sdn will remain fully supported.

### CODE REVIEW

File: pkg/types/validation/installconfig.go
Comment (danwinship) [inline pkg/types/validation/installconfig.go:337]: /hold

We're probably going to need some way to let customers install openshift-sdn with a support exception. Maybe by setting an environment variable (like with `OPENSHIFT_INSTALL_EXPERIMENTAL_DUAL_STACK`). Waiting to get some clarifications from PM...

Comment (patrickdillon) [reply pkg/types/validation/installconfig.go:337]: @danwinship The best way to allow that exception with no baggage in the installer is to tell users to run `create manifests` and then edit the network manifest. The installer will not validate user-supplied manifests. That is an extra step for users, so I'm not sure if it satisfactory, but it would work with this PR.

Comment (danwinship) [reply pkg/types/validation/installconfig.go:343]: Hm... ok, so there are 3 phases to this. (Sorry, this is still getting figured out on our side):
1. In 4.15, most customers should be blocked from creating openshift-sdn clusters, but (a) there needs to be some back door for customers with support exceptions, and (b) there needs to be a back door for the various `sdn` jobs in CI.
2. In 4.16, _all_ customers should be blocked from creating openshift-sdn clusters, and the back door that worked in 4.15 should not work any more. But CI still needs to be able to create openshift-sdn clusters, because we will still support openshift-sdn in 4.16 if you upgraded from 4.15.
3. In 4.17 openshift-sdn will be totally gone and nobody needs to be able to install new openshift-sdn clusters any more.

The "create manifests and then edit them" idea sounds like a good idea for the CI back door, although I'm not sure it's actually possible with our existing CI setup. (It seems like you can have a step that provides files to be copied into `manifests/`, but you can't have a step that runs arbitrary code in between "`create manifests`" and "`create cluster`"). (Though it's possible that every sdn-based CI job ends up with an identical `manifests/cluster-network-02-config.yml` anyway and so we could just unconditionally overwrite it?)

But that doesn't work as well for the customer back door, because of the constraint that the 4.15 solution is supposed to _not_ work in 4.16.

Comment (zaneb) [reply pkg/types/validation/installconfig.go:343]: > * In 4.15, most customers should be blocked from creating openshift-sdn clusters, but (a) there needs to be some back door for customers with support exceptions, and (b) there needs to be a back door for the various `sdn` jobs in CI.

I agree with Patrick that generating the manifests and editing them is the right way here (and conveniently already exists).

> * In 4.16, _all_ customers should be blocked from creating openshift-sdn clusters, and the back door that worked for support exceptions in 4.15 should not work any more.

There are _so_ many things you _can_ do (i.e. are not prevented by a validation) that nevertheless aren't supported that I'm not sure this one is special. In general we do a lot of validation on the install-config but little to none on any manifests that users edit.

If you are doing a thing because you have a support exception for it and that support exception expires, then if you keep doing it you will no longer be supported. This shouldn't come as a surprise to anybody that has an explicit support exception.

> But CI still needs to be able to create openshift-sdn clusters, because we will still support openshift-sdn in 4.16 if you upgraded from 4.15.

Shouldn't CI be installing 4.15 clusters and upgrading them to 4.16 to test openshift-sdn if that's the only scenario we care about?

Comment (danwinship) [reply pkg/types/validation/installconfig.go:343]: OK, actually, here's the plan (OK'ed by PM):
1. In 4.15, we drop `OpenShiftSDN` support from `install-config.yaml` as in this PR, and we will tell users with support exceptions that they can work around it by creating a manifest for `network.config.openshift.io`.
2. In 4.16, we will make CNO refuse to copy the `network.config.openshift.io` config over to a new `network.operator.openshift.io` config if the network type is `OpenShiftSDN`, which means the manifest hack we documented for 4.15 will no longer work. (It _will_ still work if you create manifests for _both_ the `network.config.openshift.io` and the `network.operator.openshift.io` objects rather than having CNO construct the operator config from the config config, and this is what CI will do, but this won't ever have been documented as something customers should do.)

File: pkg/types/validation/installconfig.go
Comment (danwinship) [inline pkg/types/validation/installconfig.go:343]: There's no reason to keep the separate error for SNO here: just use the "deprecated" message unconditionally.

Comment (danwinship) [inline pkg/types/validation/installconfig.go:343]: There are other `OpenShiftSDN` checks that can go away too. Earlier in this file there's a check to not allow ipv6 when using openshift-sdn, and elsewhere there's a special case to match `networkType: OpenShiftSDN` case-insensitively for backward compatibility with very very very old install-configs.

You should probably just grep for `OpenShiftSDN`... of course, none of those need to be fixed just to remove sdn as an install option, but you'll want to clean them up eventually.

**Review by danwinship — APPROVED** [review]

File: pkg/types/validation/installconfig.go
Comment (danwinship) [reply pkg/types/validation/installconfig.go:343]: (github hadn't updated so I didn't see your comment when I replied before.)

> There are _so_ many things you _can_ do (i.e. are not prevented by a validation) that nevertheless aren't supported that I'm not sure this one is special.

The thing that's special about this one is the fact that we will have actually _told_ people to do it, so it feels more like "I'm using a special case" than "I'm cheating". But anyway, we can block that special case from outside of the installer like I said above, so we're good here.

> Shouldn't CI be installing 4.15 clusters and upgrading them to 4.16 to test openshift-sdn if that's the only scenario we care about?

:man_shrugging: OCP in general assumes that that it can treat those two things as equivalent (as long as the 4.x-to-4.x+1 upgrade jobs are passing), and it's so much faster ( / cheaper) to _not_ test it that way.

**Comment by danwinship** [comment]
/hold

we shouldn't merge this until 4.15/master CI is ready for it... (https://github.com/openshift/release/pull/46156)

**Comment by danwinship** [comment]
> Change tests to work with newly decapitate field OpenShiftSDN

I think you mean "deprecated"?

FYI, https://github.com/openshift/release/pull/46156 has merged, so all 4.15+ aws/gcp/azure/vsphere/openstack sdn jobs should now be using the manifest hack to install sdn.

https://github.com/openshift-metal3/dev-scripts/pull/1614 has also merged, so any metal CI jobs that explicitly request `NETWORK_TYPE=OpenShiftSDN` will also use the manifest hack.

That leaves https://github.com/openshift/release/pull/46563, which fixes the last remaining metal jobs that were getting sdn by default to explicitly request it instead, so they'll get the dev-scripts fix. At that point, I _think_ this PR can be merged without breaking any jobs. FWIW, if that last PR doesn't merge soon, and you merged this PR anyway, it won't break _much_ (mostly just optional jobs anyway).

File: pkg/types/conversion/installconfig.go
Comment (zaneb) [inline pkg/types/conversion/installconfig.go:102]: I think we should keep this, because later on we are comparing against `operv1.NetworkTypeOpenShiftSDN` (e.g. in `validateNetworking()`).
Removing this conversion would make it behave like a third-party plugin (e.g. Calico) if the capitalisation didn't match exactly.

**Comment by patrickdillon** [comment]
/approve

LGTM pending fix of zane's [comment](https://github.com/openshift/installer/pull/7720#discussion_r1427394853)

**Comment by zaneb** [comment]
/lgtm
We'll need to figure out why the agent e2e test is failing and fix that in the releases repo.

**Comment by bfournie** [comment]
It looks like the required agent-compact-ipv4 required test is failing with
```
level=error msg=failed to write asset (Agent Installer ISO) to disk: cannot generate ISO image due to configuration errors
level=fatal msg=failed to fetch Agent Installer ISO: failed to load asset "Install Config": invalid install-config configuration: networking.networkType: Invalid value: "OpenShiftSDN": networkType OpenShiftSDN is deprecated, please use OVNKubernetes
```
Because of this default setting in dev-scripts https://github.com/bfournie/dev-scripts/blob/master/network.sh#L122

Will either need to override that test or change dev-scripts.

**Comment by danwinship** [comment]
ah, I missed the fact that

```
      DEVSCRIPTS_CONFIG: |
        AGENT_E2E_TEST_SCENARIO=COMPACT_IPV4
```

will result in `IP_STACK=v4` getting added...

I would suggest adding an explicit `NETWORK_TYPE=OpenShiftSDN` to `ci-operator/step-registry/agent/e2e/compact/ipv4/agent-e2e-compact-ipv4-workflow.yaml` (and then eventually fix it to be `OVNKubernetes` instead).

**Comment by bfournie** [comment]
>> (EDIT: adding NETWORK_TYPE will fix it because dev-scripts will hack things up properly if the config explicitly >> requests sdn.)

I don't think that hack at https://github.com/openshift-metal3/dev-scripts/blob/master/utils.sh#L164 is going to work properly for the agent-installer. Yes, it will get past this failure (I've confirmed that) but it won't actually set `OpenshiftSDN` in the cluster since the agent-installer isn't passing the yaml files in `manifest/`.
That being said, I'm not sure its necessary for this test to use OpenshiftSDN, it seems we can switch OVNKubernetes at this point. @zaneb wdyt?  
If not we can just set `NETWORK_TYPE=OpenShiftSDN` in ci-operator/step-registry/agent/e2e/compact/ipv4/agent-e2e-compact-ipv4-workflow.yaml at this point.

**Comment by zaneb** [comment]
I think https://github.com/openshift-metal3/dev-scripts/pull/1622 will do what we want for 4.15+ without changing the scenario on all the previous branches.
Or we could add a bit similar to https://github.com/openshift-metal3/dev-scripts/pull/1614/files#diff-35a54395325de0227bbdd613dbf61c3265b5f6ab205cf32348e3538399cf1b9eR161-R165 in the cluster-manifests step for the agent installer. @danwinship do you think that's a scenario we need to continue to test?

**Comment by zaneb** [comment]
/hold
This won't pass until we fix the agent job.

**Comment by danwinship** [comment]
> That being said, I'm not sure its necessary for this test to use OpenshiftSDN, it seems we can switch OVNKubernetes at this point.

In my initial PRs for this I was trying to avoid migrating any CI from SDN to OVN, just so I didn't have to ask that question, and then there are going to be a bunch of jira epics for 4.16 pointing out to teams where they have sdn-based CI and letting them figure out what to do with it.

For this job specifically, it seems to me like if OCP still supports using the agent installer to install openshift-sdn in 4.15, then we ought to have a CI job testing that. But I'm not sure we do... we decided we won't support the assisted installer for sdn installs any more, and it looks like the agent installer is sort of a subset of that?

**Comment by bfournie** [comment]
> For this job specifically, it seems to me like if OCP still supports using the agent installer to install openshift-sdn in 4.15, then we ought to have a CI job testing that. But I'm not sure we do... we decided we won't support the assisted installer for sdn installs any more, and it looks like the agent installer is sort of a subset of that?

Yes the agent-installer uses the assisted installer, and SDN is being removed in the assisted-installer in https://issues.redhat.com//browse/MGMT-16330

**Comment by bfournie** [comment]
Posted https://github.com/openshift/release/pull/46990

**Comment by danwinship** [comment]
https://github.com/openshift/release/pull/46990 missed some of the agent tests... filed https://github.com/openshift/release/pull/47503

**Comment by danwinship** [comment]
/hold
to avoid broken retests

File: pkg/types/validation/installconfig.go
Comment (gnufied) [inline pkg/types/validation/installconfig.go:366]: Isn't `deprecated` by definition means - still supported but support will be removed in future versions. This seems like wrong wording. It should say `Support for OpenshiftSDN is removed`.

Comment (patrickdillon) [reply pkg/types/validation/installconfig.go:366]: I think you are correct and it's a good point! I think to be on the safe side we should say something along the lines that is not supported at install time.

**Pull request merged**
